File: dynamic/best_sum.py
# 0(m^2 * n)
# 0(m^2) m being the num of nums

def best_sum(target, nums):
    table = [None] * (target + 1)
    # seed value
    table[0] = []
    for i in range(target - 1):
        if table[i] != None: 
            for num in nums:
                if (i + num) <= len(table) - 1: 
                    if table[num + i] != None: 
                        # if there is a not None value it will be equal or shorter
                        # and so there's no reason to add 
                        continue
                    else: 
                        table[num + i] = [*table[i], num]
    return table[target]

print(best_sum(100, [1, 2, 5, 25]))

File: dynamic/bestSum.js
const bestSum = (target, nums, memo = {}) => {
  if (target in memo) return memo[target];
  if (target === 0) return [];
  if (target < 0) return null;

  let shortestPath = null;

  for (let num of nums) {
    const diff = target - num;
    const remainder = bestSum(diff, nums, memo);
    if (remainder !== null) {
      const combination = [...remainder, num];
      if (shortestPath === null || combination.length < shortestPath.length) {
        shortestPath = combination;
      }
    }
  }
  memo[target] = shortestPath;
  return shortestPath;
};

console.log(bestSum(100, [1, 2, 5, 25]));
